src/foam/nanos/geocode/GoogleMapsGeocodingDAO.java

```diff
@@ -0,0 +1,168 @@
+
+public class GoogleMapsGeocodingDAO
+    extends ProxyDAO
+{
+  public static String API_HOST = "https://maps.googleapis.com/maps/api/geocode/json?address=";
+
+  protected String apiKey_;
+  protected JSONParser parser_;
+  protected ThreadLocal<StringBuilder> sb = new ThreadLocal<StringBuilder>() {
+    @Override
+    protected StringBuilder initialValue() {
+      return new StringBuilder();
+    }
+
+    @Override
+    public StringBuilder get() {
+      StringBuilder b = super.get();
+      b.setLength(0);
+      return b;
+    }
+  };
+
+  public GoogleMapsGeocodingDAO(X x, String apiKey, DAO delegate) {
+    setX(x);
+    setDelegate(delegate);
+    this.apiKey_ = apiKey;
+    this.parser_ = getX().create(JSONParser.class);
+  }
+
+  @Override
+  public FObject put_(X x, FObject obj) {
+    // don't geocode if not instance of user
+    if ( !(obj instanceof User) ) {
+      return super.put_(x, obj);
+    }
+
+    // don't geocode if no address property
+    Address address = ((User) obj).getAddress();
+    if ( address == null ) {
+      return super.put_(x, obj);
+    }
+
+    // check if address updated
+    if ( address.getLatitude() != 0 && address.getLongitude() != 0 ) {
+      User stored = (User) getDelegate().find(((User) obj).getId());
+      if (stored != null && stored.getAddress() != null) {
+        Address storedAddress = stored.getAddress();
+        // compare fields that are used to populate Google maps query
+        if ( SafetyUtil.compare(address.getAddress(), storedAddress.getAddress()) == 0 &&
+            SafetyUtil.compare(address.getCity(), storedAddress.getCity()) == 0 &&
+            SafetyUtil.compare(address.getRegionId(), storedAddress.getRegionId()) == 0 &&
+            SafetyUtil.compare(address.getPostalCode(), storedAddress.getPostalCode()) == 0 &&
+            SafetyUtil.compare(address.getCountryId(), storedAddress.getCountryId()) == 0 ) {
+          return super.put_(x, obj);
+        }
+      }
+    }
+
+    StringBuilder builder = sb.get().append(API_HOST);
+    // append address
+    if ( ! SafetyUtil.isEmpty(address.getAddress()) ) {
+      builder.append(address.getAddress()).append(",");
+    }
+
+    // append city
+    if ( ! SafetyUtil.isEmpty(address.getCity()) ) {
+      builder.append(address.getCity()).append(",");
+    }
+
+    // append province
+    if ( ! SafetyUtil.isEmpty((String) address.getRegionId()) ) {
+      builder.append((String) address.getRegionId()).append(",");
+    }
+
+    // append postal code
+    if ( ! SafetyUtil.isEmpty(address.getPostalCode()) ) {
+      builder.append(address.getPostalCode()).append(",");
+    }
+
+    // append country id
+    if ( ! SafetyUtil.isEmpty((String) address.getCountryId()) ) {
+      builder.append((String) address.getCountryId());
+    }
+
+    // append api key
+    builder.append("&key=").append(apiKey_);
+
+    String line = null;
+    HttpURLConnection conn = null;
+    BufferedReader reader = null;
+
+    try {
```

Comment:
Would it make sense for this to be a separate GeoService in the Context rather than being in-lined here? In case you wanted to GeoLocate in some other context?

Reply:
Yeah it would make sense. I can do that in a following PR.